test/zdtm/static/autofs.c

```diff
@@ -63,7 +63,6 @@ static char *xvstrcat(char *str, const char *fmt, va_list args)
 	if (ret == -ENOMEM) {
 		/* realloc failed. We must release former string */
 		pr_err("Failed to allocate string\n");
-		free(str);
```

Comment:
I believe that with this you break the underline logic of xvstrcat that it should free str on error, imagine that on first iteration of do-while loop realloc fails, str is still valid and we need to free it, because the caller will not know if it was fail on first iteration or on the second where str is already freed implicitly by realloc.

I believe we should instead do:
```
diff --git a/test/zdtm/static/autofs.c b/test/zdtm/static/autofs.c
index 53ece4458..ad1795842 100644
--- a/test/zdtm/static/autofs.c
+++ b/test/zdtm/static/autofs.c
@@ -47,6 +47,7 @@ static char *xvstrcat(char *str, const char *fmt, va_list args)
                ret = -ENOMEM;
                new = realloc(str, offset + delta);
                if (new) {
+                       str = new;
                        va_copy(tmp, args);
                        ret = vsnprintf(new + offset, delta, fmt, tmp);
                        va_end(tmp);
@@ -54,7 +55,6 @@ static char *xvstrcat(char *str, const char *fmt, va_list args)
                                /* NOTE: vsnprintf returns the amount of bytes
                                 *                                  * to allocate. */
                                delta = ret + 1;
-                               str = new;
                                ret = 0;
                        }
                }
```

Now str is always updated after successful realoc, not conditionally as it was before, so it always represent the currently allocated string.

Reply:
Thanks @Snorch!
I moved this patch in https://github.com/checkpoint-restore/criu/pull/1732/ and applied this change there.